File: packages/sdk/src/wallet/challenge.ts
import { randomBytes } from 'crypto';
import type { ChallengeMessage } from '@eterecitizen/common';

const CHALLENGE_EXPIRY_MS = 5 * 60 * 1000; // 5 minutes

export function generateChallenge(address: string, chainId: number): ChallengeMessage {
  const nonce = randomBytes(16).toString('hex');
  const now = new Date();

  return {
    domain: 'eterecitizen.ai',
    address,
    statement: 'Sign this message to prove ownership of this wallet for EtereCitizen protocol.',
    nonce,
    issuedAt: now.toISOString(),
    expiresAt: new Date(now.getTime() + CHALLENGE_EXPIRY_MS).toISOString(),
    chainId,
  };
}

export function challengeToMessage(challenge: ChallengeMessage): string {
  return [
    `${challenge.domain} wants you to sign in with your Ethereum account:`,
    challenge.address,
    '',
    challenge.statement,
    '',
    `Chain ID: ${challenge.chainId}`,
    `Nonce: ${challenge.nonce}`,
    `Issued At: ${challenge.issuedAt}`,
    `Expiration Time: ${challenge.expiresAt}`,
  ].join('\n');
}

export function isChallengeExpired(challenge: ChallengeMessage): boolean {
  return new Date(challenge.expiresAt).getTime() < Date.now();
}

File: sdks/python/eterecitizen/challenge.py
"""Wallet-ownership challenge serialization. See SPEC §12.1.2."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ChallengeMessage:
    domain: str
    address: str
    statement: str
    nonce: str
    issuedAt: str   # ISO 8601
    expiresAt: str  # ISO 8601
    chainId: int


def challenge_to_message(challenge: ChallengeMessage | dict) -> str:
    """Canonical serialization. Lines are separated by LF (\\n). Blank lines
    separate the header, statement, and parameter block. SPEC §12.1.2.
    """
    if isinstance(challenge, dict):
        c = challenge
    else:
        c = {
            "domain": challenge.domain,
            "address": challenge.address,
            "statement": challenge.statement,
            "nonce": challenge.nonce,
            "issuedAt": challenge.issuedAt,
            "expiresAt": challenge.expiresAt,
            "chainId": challenge.chainId,
        }

    return "\n".join([
        f"{c['domain']} wants you to sign in with your Ethereum account:",
        c["address"],
        "",
        c["statement"],
        "",
        f"Chain ID: {c['chainId']}",
        f"Nonce: {c['nonce']}",
        f"Issued At: {c['issuedAt']}",
        f"Expiration Time: {c['expiresAt']}",
    ])
